compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/AMD64ControlFlow.java

```diff
@@ -314,6 +314,97 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
         }
     }
 
+    public static final class HashTableSwitchOp extends AMD64BlockEndOp {
+        public static final LIRInstructionClass<HashTableSwitchOp> TYPE = LIRInstructionClass.create(HashTableSwitchOp.class);
+        private final JavaConstant[] keys;
+        private final LabelRef defaultTarget;
+        private final LabelRef[] targets;
+        @Use protected Value hash;
+        @Temp({REG, HINT}) protected Value idxScratch;
+        @Temp({REG, HINT}) protected Value valueScratch;
+        @Temp protected Value scratch;
+        @Use protected Value value;
+
+        public HashTableSwitchOp(final JavaConstant[] keys, final LabelRef defaultTarget, LabelRef[] targets, Value value, Value hash, Variable scratch, Variable idxScratch, Variable valueScratch) {
+            super(TYPE);
+            this.keys = keys;
+            this.defaultTarget = defaultTarget;
+            this.targets = targets;
+            this.value = value;
+            this.hash = hash;
+            this.scratch = scratch;
+            this.idxScratch = idxScratch;
+            this.valueScratch = valueScratch;
+        }
+
+        @Override
+        public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
+            Register valueReg = asRegister(value, AMD64Kind.DWORD);
+            Register indexReg = asRegister(hash, AMD64Kind.DWORD);
+            Register idxScratchReg = asRegister(idxScratch, AMD64Kind.DWORD);
+            Register valueScratchReg = asRegister(valueScratch, AMD64Kind.DWORD);
+            Register scratchReg = asRegister(scratch, AMD64Kind.QWORD);
+
+            if (!indexReg.equals(idxScratchReg)) {
+                masm.movl(idxScratchReg, indexReg);
+            }
+
+            // Set scratch to address of jump table
+            masm.leaq(scratchReg, new AMD64Address(AMD64.rip, 0));
+            final int afterLea = masm.position();
+
+            // Jump to default target if index doesn't contain the expected value
+            if (defaultTarget != null) {
+                masm.movslq(valueScratchReg, new AMD64Address(scratchReg, idxScratchReg, Scale.Times8, 0));
```

Comment:
you might consider loading 8 bytes here instead of issuing two loads.  If you put key is in the low 32 bits you can still use cmpl and then just perform a shift instead of the second load.  You'd want to align the table at 8 bytes instead which might be a good idea in any case when you have both key and offset in the table.  Having a single load would also allow you to eliminate the move of indexReg to idxScratchReg since you'd only have a single use of it.